sssd_test_framework/topology_controllers.py

```diff
@@ -14,59 +23,173 @@
+            if not isinstance(host, BaseBackupHost):
+                continue
+
+            restore_data[host] = host.backup_data
+
+        self.restore(restore_data)
+
+    def topology_teardown(self) -> None:
+        if self.provisioned:
+            return
+
+        try:
+            for host, backup_data in self.backup_data.items():
+                if not isinstance(host, BaseBackupHost):
+                    continue
+
+                host.remove_backup(backup_data)
+        except Exception:
+            # This is not that important, we can just ignore
+            pass
+
+        self.restore_vanilla()
+
+    def teardown(self) -> None:
+        if self.provisioned:
+            self.restore_vanilla()
+            return
+
+        self.restore(self.backup_data)
+
 
 class ClientTopologyController(BackupTopologyController):
     """
     Client Topology Controller.
     """
 
-    pass
+    def topology_teardown(self) -> None:
+        pass
+
+    def teardown(self) -> None:
+        self.restore_vanilla()
 
 
 class LDAPTopologyController(BackupTopologyController):
     """
     LDAP Topology Controller.
     """
 
-    pass
+    def topology_teardown(self) -> None:
+        pass
+
+    def teardown(self) -> None:
+        self.restore_vanilla()
 
 
 class IPATopologyController(BackupTopologyController):
     """
     IPA Topology Controller.
     """
 
-    pass
+    @restore_vanilla_on_error
+    def topology_setup(self, client: ClientHost, ipa: IPAHost, nfs: NFSHost) -> None:
+        if self.provisioned:
+            self.logger.info(f"Topology '{self.name}' is already provisioned")
+            return
+
+        self.logger.info(f"Enrolling {client.hostname} into {ipa.domain}")
+
+        # Remove any existing Kerberos configuration and keytab
+        client.fs.rm("/etc/krb5.conf")
+        client.fs.rm("/etc/krb5.keytab")
+
+        # Backup ipa-client-install files
+        client.fs.backup("/etc/ipa")
+        client.fs.backup("/var/lib/ipa-client")
+
+        # Join ipa domain
+        client.ssh.exec(["realm", "join", ipa.domain], input=ipa.adminpw)
+
+        # Backup so we can restore to this state after each test
+        self.backup_data[ipa] = ipa.backup()
+        self.backup_data[client] = client.backup()
+        self.backup_data[nfs] = nfs.backup()
 
 
 class ADTopologyController(BackupTopologyController):
     """
     AD Topology Controller.
     """
 
-    pass
+    @restore_vanilla_on_error
+    def topology_setup(self, client: ClientHost, provider: ADHost | SambaHost, nfs: NFSHost) -> None:
+        if self.provisioned:
+            self.logger.info(f"Topology '{self.name}' is already provisioned")
+            return
 
+        self.logger.info(f"Enrolling {client.hostname} into {provider.domain}")
 
-class SambaTopologyController(BackupTopologyController):
+        # Remove any existing Kerberos configuration and keytab
+        client.fs.rm("/etc/krb5.conf")
+        client.fs.rm("/etc/krb5.keytab")
```

Comment:
I'm a bit unclear why this is needed if realm join updates these files.

Reply:
There may be already some principal stored in the keytab or some existing kerberos configuration. In theory, we could just backup these files so we can revert any changes, but imho removing them is better so we always start with a clean state.